### ISSUE TITLE
Agate is not listening on multiple addresses

### ISSUE THREAD

**Issue by alvaro-cuesta** [mbrubeck/agate#63]
_(Related to #14)_

According to [this line](https://github.com/mbrubeck/agate/blob/ba3c01d38f85c11326f365a6f6be780e698afca4/src/main.rs#L103-L108) in the program help, when multiple bind addresses are specified Agate will listen on all of them.

Addresses are not handled individually, and [are just passed down to `TcpListener::bind`](https://github.com/mbrubeck/agate/blob/ba3c01d38f85c11326f365a6f6be780e698afca4/src/main.rs#L46), so this means that Agate will actually NOT listen on multiple addresses, only on the first that it can bind to.

See [`TcpListener::bind` docs](https://docs.rs/tokio/1.8.1/tokio/net/struct.TcpListener.html#method.bind):

> If `addr` yields multiple addresses, bind will be attempted with each of the addresses until one succeeds and returns the listener.

This can be confirmed on Windows, where "dual stack" IPv6 is not a default (at least on Rust). The default Agate config will NOT respond on 127.0.0.1, only on [::1] (or on localhost, which points to both).

---

I ran into this issue when programming my own toy Gemini server in Rust and was just snooping on how you did it 😝

I haven't succeeded on binding to multiple addresses so far (not even via [`socket2` crate](https://docs.rs/socket2/0.4.0/socket2/) as suggested by Tokio's docs) since a socket can only be either IPv6 or IPv4, and I'd want to bind it cross-domain... so I'm starting to think we really need multiple sockets, one per address.

Dual stack can probably be enabled via [`.only_v6()`](https://docs.rs/socket2/0.4.0/socket2/struct.Socket.html#method.only_v6) but it still means dealing with IPv4 and IPv6 addresses separately due to the cross-domain issue.

I will keep you updated if I make any progress.

**Comment by Johann150** [comment]
😓 I better read the documentation next time...

I think the easiest solution I'll go for is to start a "server" (thread) for each IP to listen on. Tokio's `select!` also looks interesting but since that is a macro I don't think it can deal with a changing number of things, so it can not be used in this case.

**Comment by alvaro-cuesta** [comment]
> 😓 I better read the documentation next time...

If it's any consolation I did exactly the same. I was just lucky to be on Windows and it nudged me into the right direction 😝 

> Tokio's select! also looks interesting

Yup, been thinking about that too.

One thread per socket doesn't seem to be much of a waste, but I found [`futures::future::select_all`](https://docs.rs/futures/0.3.15/futures/future/fn.select_all.html). Is this compatible with Tokio's futures?
